### PR TITLE
Refactor pypit_setup

### PR DESCRIPTION
**Pull request by profxj**
This PR makes a pass through all of the code used to 
run the pypit_setup script.   My main intent was to further
modularize the code, where sensible.  This primarily meant
breaking apart a few of the longest methods.  I also pushed
a set of key methods into a new module:  arsetup.py

The only fundamental change that was started but not
fully implemented was to generate a new object cal_indx
that I have in mind to replace the functionality of 
settings.spect[frametype]['index']

I should like to advocate that settings.spect (and argf) be
immutable.  That is, they are set at the start and then
guide the reduction process.  While it is convenient that they are
globally accessible, I should prefer that they not be used
as a means to pass new information generated by the code.
Happy to discuss further, especially with @rcooke-ast .

Documentation was expanded in support of setup.
Additional tests were added.

All tests in the Development-Suite are passing.

### CODE REVIEW

**Comment by profxj** [comment]
Setup files are now put in a generated setup_files/ folder.
Documentation has been updated in the Cookbook.
Addresses Issue #288

**Comment by rcooke-ast** [comment]
Don't forget to add a note in CHANGES.rst!  :-)

(a review of the code is coming shortly...)

File: doc/internals.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+setup_dict
+----------
+
+This highly nested *dict* organizes the various setups used in the input set of
+data files.  The top-level keys, which define the setup,
+are simple labels:  A, B, C, ...
+
+The next set of keys are:
+
+*  '--': which holds a *dict* defining properties of the setup (e.g. dichroic)
```
[the remark this answers is not in the document]
Comment (profxj) [reply doc/internals.rst]: Have added a partial example.

File: pypit/arload.py [hunk 2]
```diff
@@ -142,7 +147,7 @@ def generate_updates(dct, keylst, keys, whddict, headarr):
                     except KeyError: # Keyword not found in header
                         msgs.warn("{:s} keyword not in header. Setting to None".format(kchk))
                         value=str('None')
-            # Convert the input time into hours
+            # Convert the input time into hours -- Should we really do this here??
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/arload.py]: I suppose I should prefer MJD but I'm ok
with this for now.

File: pypit/armbase.py [hunk 3]
```diff
@@ -44,41 +43,39 @@ def SetupScience(fitsdict):
...
-        setup_dict = {}
-        setup_dict[setup[0]] = {}
-        for ii in range(1,20): # Dummy detectors
-            setup_dict[setup[0]][arparse.get_dnum(ii)] = dict(binning='1x1')
-        setup_dict[setup[0]][setup[-2:]] = {}
-        iSCI = filesort['science']
-        setup_dict[setup[0]][setup[-2:]]['sci'] = [fitsdict['filename'][i] for i in iSCI]
-        # Write
-        calib_file = arsort.write_calib(setup_dict)
+            msgs.error("Need to specify --  reduce masters setup  -- in your PYPIT file!")
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/armbase.py]: Good suggestion. Modified.

File: pypit/arsetup.py [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+    """ Generates a dummy setup_dict
+
+    Parameters
+    ----------
+    filesort : dict
+    fitsdict : dict
+
+    Returns
+    -------
+    setup_dict : dicg
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/arsetup.py]: fixed

File: pypit/arsetup.py [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+      detector dict
+
+    Returns
+    -------
+    dkey : str
+      Name of the detector string associated to the input ddict
+      May be new or previously used
+
+    """
+    det_str = [str('{:02d}'.format(i+1)) for i in range(99)]
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/arsetup.py]: good suggestion.  Modified (with prefix=False)

**Comment by profxj** [comment]
Looks all clear now.

Merging.

**Pull request merged**